### PR TITLE
Feat/6 ledger accessor mutator functions

### PR DESCRIPTION
**Pull request by AlexTheWizardL**
# Pull Request: ledger accessor mutator functions

## Overview

This pull request introduces significant enhancements to the ledger management functionalities within the CashCtrl integration. The following key features and improvements are included:

1. **Enhanced `ledger` Method**:
   - Retrieves and formats the account remote ledger from CashCtrl to the pyledger schema.
   - Supports both collective and single journal entries.
   - Maps account IDs and tax rates for better readability and accuracy.

2. **New `add_ledger_entry` Method**:
   - Allows adding new ledger entries to the remote CashCtrl instance.
   - Supports both single and collective transactions.
   - Includes validation to ensure a single currency is used in collective bookings.

3. **New `update_ledger_entry` Method**:
   - Allows updating existing ledger entries in the remote CashCtrl instance.
   - Supports both single and collective transactions.
   - Ensures consistency and integrity of the updated data.

4. **New `delete_ledger_entry` Method**:
   - Allows deleting ledger entries from the remote CashCtrl instance.
   - Supports deletion of multiple entries at once.

5. **Unit Tests**:
   - Introduces unit tests using pytest to verify the functionality of the ledger mutator methods.
   - Includes tests for adding, updating, and deleting ledger entries.
   - Ensures accurate VAT code handling and data consistency.

### Unit Tests

- Adds tests for adding, updating, and deleting ledger entries.
- Uses pytest fixtures for setting up VAT codes.
- Verifies data consistency and correctness.

@lasuk Please review the changes and provide feedback or approval for merging.

### CODE REVIEW

**Comment by lasuk** [comment]
I've re-factored the `ledger()` method in 0a8b163 with the goal to enhance code readability and to explain individual vs. collective bookings.

@AlexTheWizardL: Please shout if you disagree with any changes.

Btw: This is a good example of how passing (not-routinely failing) tests help in our collaboration. The green check-mark indicates that re-factoring was successful (i.e. did not break the code or alter the outcome). We can now both trust that my commit did not break your code. Instead of scrutinising the code for potential errors, we can focus on the big picture and judge whether the code changes contribute to improving our code base in terms of readability, etc.

**Review by lasuk — COMMENTED** [review]
Thanks for implementing the ledger accessor and mutator functions.

I have added a series of comments. Please address or discuss further.

**Comment by lasuk** [comment]
I am thinking about changing the interface of `add_ledger_entry()` and `update_ledger_entry()` to pass directly a ledger data frame, rather than passing 'date' and 'id' as separate functions. With the proposed interface, the method would raise an error if 'date' or 'id' are not unique. The proposed change has the advantage of reducing the number of data base schemas used in pyledger. What do you think @AlexTheWizardL ?

Current interface:
```python
add_ledger_entry(self, date: datetime.date, target: pd.DataFrame)
update_ledger_entry(self, id: str, date: datetime.date, target: pd.DataFrame)
# 'entry' is a DataFrame with LedgerEngine.ledger() schema minus columns 'date' and 'id'.
```

Proposed interface:
```python
add_ledger_entry(self, entry: pd.DataFrame)
update_ledger_entry(self, entry: pd.DataFrame)
# 'entry' is a DataFrame with LedgerEngine.ledger() schema.
```


I know this suggestion comes a little late in the process. Let's discuss the proposal and the best way forward in the call.

File: cashctrl_ledger/constants.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
+"""
+This module contains constants used throughout the application.
+"""
+
+JOURNAL_SUB_ROWS_COLUMNS = {
```
[the remark this answers is not in the document]
Comment (AlexTheWizardL) [reply cashctrl_ledger/constants.py:5]: I like your idea, here are some suggested names:
`JOURNAL_ITEM_COLUMNS`
`COLLECTIVE_COLUMNS`
`COLLECTIVE_JOURNAL_COLUMNS`
`LEDGER_ITEM_COLUMNS`

[the remark this answers is not in the document]
Comment (lasuk) [reply cashctrl_ledger/constants.py:5]: Thanks. I vote for `COLLECTIVE_JOURNAL_COLUMNS` or `JOURNAL_TRANSACTION_COLUMNS`.

**Comment by lasuk** [comment]
I've added two commits
- 6c1f983 with minor corrections to prettify the code
- 9cc1dfc to expand and streamline test logic:
    - Avoid fixtures where there is no benefit in the execution logic.
    - More robust argument handling in `add_ledger_entry()` and `update_ledger_entry()` (both methods standardize the given data frame).
    - Test replacing individual with collective entries and vice versa.

@AlexTheWizardL: Please shout if you disagree.

File: tests/test_ledger.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+    # initial_ledger = cashctrl_ledger.ledger().reset_index(drop=True)
+    # new_entry = individual_transaction.copy()
+    # new_entry['vat_code'] = None
+    # new_entry['text'] = 'Ledger entry without VAT'
+    # cashctrl_ledger.add_ledger_entry(entry=new_entry)
+    # updated_ledger = cashctrl_ledger.ledger().reset_index(drop=True)
+    # outer_join = pd.merge(initial_ledger, updated_ledger, how='outer', indicator=True)
+    # created = outer_join[outer_join['_merge'] == "right_only"].drop('_merge', axis = 1).reset_index(drop=True)
+    # expected = StandaloneLedger.standardize_ledger(new_entry)
+    # pd.testing.assert_frame_equal(created.drop(columns=['id']), expected.drop(columns=['id']))
```
Comment (lasuk) [inline tests/test_ledger.py:70]: Tests fail. Please uncomment and fix.

File: tests/test_ledger.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+    # initial_ledger = cashctrl_ledger.ledger().reset_index(drop=True)
+    # new_entry = collective_transaction.copy()
+    # new_entry['id'] = created.at[0, 'id']
+    # new_entry.at[0, 'vat_code'] = None
+    # cashctrl_ledger.update_ledger_entry(entry=new_entry)
+    # updated_ledger = cashctrl_ledger.ledger().reset_index(drop=True)
+    # outer_join = pd.merge(initial_ledger, updated_ledger, how='outer', indicator=True)
+    # updated = outer_join[outer_join['_merge'] == "right_only"].drop('_merge', axis = 1).reset_index(drop=True)
+    # expected = StandaloneLedger.standardize_ledger(new_entry)
+    # pd.testing.assert_frame_equal(updated, expected)
```
Comment (lasuk) [inline tests/test_ledger.py:94]: Test fails. Please uncomment and fix.

File: tests/test_ledger.py [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+    # # Test adding a ledger entry without VAT
+    # initial_ledger = cashctrl_ledger.ledger().reset_index(drop=True)
+    # new_entry = collective_transaction.copy()
+    # new_entry['vat_code'] = None
+    # cashctrl_ledger.add_ledger_entry(entry=new_entry)
+    # updated_ledger = cashctrl_ledger.ledger().reset_index(drop=True)
+    # outer_join = pd.merge(initial_ledger, updated_ledger, how='outer', indicator=True)
+    # created = outer_join[outer_join['_merge'] == "right_only"].drop('_merge', axis = 1).reset_index(drop=True)
+    # expected = StandaloneLedger.standardize_ledger(new_entry)
+    # pd.testing.assert_frame_equal(created.drop(columns=['id']), expected.drop(columns=['id']))
```
Comment (lasuk) [inline tests/test_ledger.py:127]: Tests fail. please uncomment and fix.

File: tests/test_ledger.py [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+    # initial_ledger = cashctrl_ledger.ledger().reset_index(drop=True)
+    # new_entry = individual_transaction.copy()
+    # new_entry['id'] = created.at[0, 'id']
+    # new_entry['vat_code'] = None
+    # cashctrl_ledger.update_ledger_entry(entry=new_entry)
+    # updated_ledger = cashctrl_ledger.ledger().reset_index(drop=True)
+    # outer_join = pd.merge(initial_ledger, updated_ledger, how='outer', indicator=True)
+    # updated = outer_join[outer_join['_merge'] == "right_only"].drop('_merge', axis = 1).reset_index(drop=True)
+    # expected = StandaloneLedger.standardize_ledger(new_entry)
+    # pd.testing.assert_frame_equal(updated, expected)
```
Comment (lasuk) [inline tests/test_ledger.py:152]: The test fails. Please uncomment and fix.

**Review by lasuk — COMMENTED** [review]
Some tests added in commit 9cc1dfc are commented out. They fail if uncommented.

Please uncomment the tests and address the issues.

**Pull request merged**